spectrum-network/src/network_controller.rs

```diff
@@ -0,0 +1,490 @@
+                }) = self.enabled_peers.get_mut(&peer_id)
+                {
+                    let protocol_id = protocol_tag.protocol_id();
+                    let (_, prot_handler) = self.supported_protocols.get(&protocol_id).unwrap();
+                    match enabled_protocols.entry(protocol_id) {
+                        Entry::Vacant(entry) => {
+                            entry.insert((EnabledProtocol::PendingApprove, prot_handler.clone()));
+                            prot_handler.protocol_requested(peer_id, protocol_tag.protocol_ver(), handshake);
+                        }
+                        Entry::Occupied(_) => {
+                            trace!(
+                                "Peer {:?} opened already enabled protocol {:?}",
+                                peer_id,
+                                protocol_id
+                            );
+                            self.pending_actions
+                                .push_back(NetworkBehaviourAction::NotifyHandler {
+                                    peer_id,
+                                    handler: NotifyHandler::One(connection),
+                                    event: ConnHandlerIn::Close(protocol_id),
+                                })
+                        }
+                    }
+                }
+            }
+            ConnHandlerOut::ClosedByPeer(protocol_id)
+            | ConnHandlerOut::RefusedToOpen(protocol_id)
+            | ConnHandlerOut::Closed(protocol_id) => {
+                if let Some(ConnectedPeer::Connected {
+                    enabled_protocols, ..
+                }) = self.enabled_peers.get_mut(&peer_id)
+                {
+                    match enabled_protocols.entry(protocol_id) {
+                        Entry::Occupied(entry) => {
+                            trace!(
+                                "Peer {:?} closed the substream for protocol {:?}",
+                                peer_id,
+                                protocol_id
+                            );
+                            entry.remove();
+                        }
+                        Entry::Vacant(_) => {}
+                    }
+                }
+            }
+            ConnHandlerOut::Message {
+                protocol_tag,
+                content,
+            } => {
+                if let Some(ConnectedPeer::Connected {
+                    enabled_protocols, ..
+                }) = self.enabled_peers.get_mut(&peer_id)
+                {
+                    let protocol_id = protocol_tag.protocol_id();
+                    match enabled_protocols.get(&protocol_id) {
+                        Some((_, prot_handler)) => {
+                            prot_handler.incoming_msg(peer_id, protocol_tag.protocol_ver(), content);
+                        }
+                        None => {} // todo: probably possible?
+                    };
+                }
+            }
+        }
+    }
+
+    fn poll(
```

Comment:
Is there a particular reason why you are implementing a `poll` function here? I notice an implementation of `Future` in `PeerManager` too. I believe that we can do away with implementing it manually and use combinators to define a  new stream. In particular we can use `futures::stream::select_all()` to merge the streams from `self.peer_manager` and `self.requests_recv` (we can define an enum to unify `NetworkControllerIn` with `PeerManagerOut`).

This simplifies the code a lot plus we automatically defer to their implementation of `Stream`, so it will properly yield when nothing is coming through etc.

Another thing we can consider is replacing `self.pending_actions` with a bounded message channel.

Sorry there's a lot in this comment. If you like I can make a PR on top of this to demonstrate

Reply:
Sorry my bad, I see that it's an impl of `NetworkBehaviour` in `libp2p`.